### PR TITLE
Add parameter metadata to the trainer schema.

### PR DESCRIPTION
**Pull request by justinxzhao**

### CODE REVIEW

**Review by connor-mccorm — APPROVED** [review]

File: ludwig/schema/optimizers.py [hunk 1]
```diff
@@ -354,7 +355,12 @@ class GradientClippingConfig(BaseMarshmallowConfig):
     clipvalue: Optional[float] = FloatRange(default=None, allow_none=True, description="")
 
 
-def GradientClippingDataclassField(default={}, allow_none=True, description="TODO"):
+def GradientClippingDataclassField(
+    default={},
+    allow_none=True,
+    description="Parameters for gradient clipping.",
+    parameter_metadata: ParameterMetadata = None,
```
Comment (ksbrar) [inline ludwig/schema/optimizers.py:362]: I would not let `parameter_metadata` be set here, and just have it set to `TRAINER_METADATA['gradient_clipping']`. Actually I'd remove all the parameters here, being only used in the trainer it's something that should not have open arguments.

File: ludwig/schema/trainer.py [hunk 2]
```diff
@@ -229,7 +310,10 @@ class GBMTrainerConfig(BaseTrainerConfig):
     eval_batch_size: int = schema_utils.PositiveInteger(
```
Comment (ksbrar) [inline ludwig/schema/trainer.py:310]: Metadata omitted on purpose? Is the `eval_batch_size` metadata necessarily different for GBM?

File: ludwig/schema/utils.py [hunk 3]
```diff
@@ -687,13 +738,20 @@ def _jsonschema_type_mapping(self):
                 else []
             )
 
-            return {"oneOf": oneof_list, "title": self.name, "description": description, "default": default}
+            return {
```
Comment (ksbrar) [inline ludwig/schema/utils.py:741]: Not part of the PR, but would you mind adding this line above after `numeric_option` is declared?

```
if not is_integer: numeric_option["format"] = "float"
```

**Review by ksbrar — CHANGES_REQUESTED** [review]

File: ludwig/schema/optimizers.py (hunk 1 above)
Comment (justinxzhao) [reply ludwig/schema/optimizers.py]: Done, though I left `description` and `default` since I think it's still visually useful for those fields to appear in the Trainer schema.

File: ludwig/schema/trainer.py (hunk 2 above)
Comment (justinxzhao) [reply ludwig/schema/trainer.py]: I don't think so, though this probably means we can just use the `BaseTrainerConfig`'s `eval_batch_size` definition.

File: ludwig/schema/utils.py (hunk 3 above)
Comment (justinxzhao) [reply ludwig/schema/utils.py:743]: Done.

File: ludwig/schema/metadata/trainer_metadata.py [hunk 4]
```diff
@@ -0,0 +1,685 @@
...
+# fmt: off
+from ludwig.schema.metadata.parameter_metadata import ParameterMetadata
+
+TRAINER_METADATA = (
+{'batch_size': ParameterMetadata(ui_display_name='Batch Size',
+                                 default_value_reasoning='Not too big, not too small.',
+                                 example_value=None,
+                                 related_parameters='eval_batch_size',
+                                 other_information=None,
+                                 description_implications="There's conflicting advice about what batch size to use. "
```
[the remark this answers is not in the document]
Comment (justinxzhao) [reply ludwig/schema/metadata/trainer_metadata.py:11]: Yes! pretty printing for dataclasses is supported in python 3.10.

File: ludwig/schema/metadata/trainer_metadata.py [hunk 5]
```diff
@@ -0,0 +1,685 @@
...
+                                                          'Using a higher batch size will achieve the highest '
+                                                          "throughput and training efficiency. However, there's also "
+                                                          'evidence that depending on other hyperparameters, a smaller '
+                                                          'batch size may produce a higher quality model.',
+                                 suggested_values='auto',
+                                 suggested_values_reasoning='Try at least a few different batch sizes to get a sense '
+                                                            'of whether batch size affects model performance',
+                                 commonly_used=True,
+                                 expected_impact='HIGH',
```
[the remark this answers is not in the document]
Comment (justinxzhao) [reply ludwig/schema/metadata/trainer_metadata.py]: Done.

File: ludwig/schema/optimizers.py [hunk 6]
```diff
@@ -354,13 +355,14 @@ class GradientClippingConfig(BaseMarshmallowConfig):
     clipvalue: Optional[float] = FloatRange(default=None, allow_none=True, description="")
 
 
-def GradientClippingDataclassField(default={}, allow_none=True, description="TODO"):
+def GradientClippingDataclassField(description: str, default: Dict):
```
Comment (ksbrar) [inline ludwig/schema/optimizers.py:358]: I would still keep `{}` as the default value for `default` so it's clear that in the case of `GradientClippingDataclassField` an empty dict is valid (whereas e.g. for `OptimizerDataclassField` you have to at least provide a `type` inside the dict)

File: ludwig/schema/trainer.py [hunk 7]
```diff
@@ -226,12 +307,6 @@ class GBMTrainerConfig(BaseTrainerConfig):
         allow_none=False,
     )
 
-    eval_batch_size: int = schema_utils.PositiveInteger(
```
Comment (ksbrar) [inline ludwig/schema/trainer.py:229]: Just want to double check, `eval_batch_size` here had the `default` set to `128` whereas the `default` on `BaseTrainerConfig` is `None`

**Review by ksbrar — CHANGES_REQUESTED** [review]

File: ludwig/schema/optimizers.py (hunk 6 above)
Comment (justinxzhao) [reply ludwig/schema/optimizers.py]: Added back the default value for `default` for now.

Outside of this PR, we should consider removing the default values for marshmallow field function arguments. This will make some of the marshmallow unit tests more verbose, but it ensures that the field's parameters (default, description, etc.) are defined in one location, at the point of instantiation. For example, right now for `OptimizerDataclassField`, we declare the `default` in two places -- as a function parameter default via the marshmallow field definition, as well as at the point of instantiation in `ECDTrainerConfig`.

Here's what I'm thinking:
- Always include `description` and `default` as required arguments so that it's impossible to instantiate a marshmallow field without explicitly specifying the `description` and `default`. This setup ensures that the defaults and descriptions are concisely listed where parameters are declared, i.e. `ECDTrainerConfig`.
- Include other arguments like `allow_none` in the marshmallow field only if we want this to be able to be configured differently across different instances of the field. For example, we should remove `allow_none` from the list of parameters for `GradientClippingDataclassField` since it's always `True`.
- Use `ParameterMetadata` to contain all other "optional" information pertaining to the field. Eventually, this will also be a required argument.

See also, [why default arguments are bad](https://quuxplusone.github.io/blog/2020/04/18/default-function-arguments-are-the-devil/).

Comment (ksbrar) [reply ludwig/schema/optimizers.py]: Agreed on all counts!

**Review by ksbrar — APPROVED** [review]
